push to stable

vanhauser-thc opened the pull request:
before we get a larger readme change from Jana maybe we should push to stable.

vanhauser-thc committed:
add grammatron custom mutator

vanhauser-thc committed:
fix compile

vanhauser-thc committed:
fixes

vanhauser-thc committed:
more fixes

vanhauser-thc committed:
reenable splicing

vanhauser-thc committed:
port custom mutator changes

vanhauser-thc committed:
3.15a init

andreafioraldi committed:
update qemuafl

andreafioraldi committed:
Merge branch 'dev' of github.com:AFLplusplus/AFLplusplus into dev

adrianherrera committed:
utils: added optimin corpus minimizer

vanhauser-thc committed:
Merge pull request #1034 from AFLplusplus/grammatron

Grammatron

vanhauser-thc committed:
add gramatron

vanhauser-thc committed:
remove realpath dependency

vanhauser-thc committed:
readd realpath

moshekaplan committed:
Update README.persistent_mode.md

vanhauser-thc committed:
Merge pull request #1036 from moshekaplan/patch-1

Update README.persistent_mode.md

vanhauser-thc committed:
fix rust bindings

Maaxxs committed:
Fix typos (#1037)

domenukk committed:
fix rust mutator bindingsbuild

domenukk committed:
fixed rust example

vanhauser-thc committed:
Merge pull request #1035 from adrianherrera/optimin-util

add optimin corpus minimizer

vanhauser-thc committed:
optimin nits

vanhauser-thc committed:
fix optimin readme

yuawn committed:
fix links in FAQ.md

yuawn committed:
remove unneeded assignment

vanhauser-thc committed:
Merge pull request #1039 from yuawn/fix-save_if_interesting

remove unneeded assignment

yuawn committed:
update relative path

vanhauser-thc committed:
Merge pull request #1038 from yuawn/fix-faq

Fix links in FAQ.md

adrianherrera committed:
optimin: add missing cmin options and auto create output dir

* Includes frida, unicorn modes
* Automatically create the output directory if it does not already exist

vanhauser-thc committed:
MacOS nits

jhertz-apple committed:
persistent SHM config install for Mac

vanhauser-thc committed:
Merge pull request #1041 from jhertz/mac_shm

persistent SHM config install for Mac

vanhauser-thc committed:
afl-persistent-config Linux support + documentation

vanhauser-thc committed:
doc: stop, restart

vanhauser-thc committed:
update grammar mutator

adrianherrera committed:
optimin: improve afl-showmap error messages

adrianherrera committed:
optimin: recursively iterate over input directory

vanhauser-thc committed:
handle single seed with perf_score 0

vanhauser-thc committed:
removed utils/afl_frida

yuawn committed:
remove redundant if statement

domenukk committed:
give shmem more space

domenukk committed:
aligned wording

adrianherrera committed:
optimin: add support for afl-cmin env vars

adrianherrera committed:
optimin: add test of binary target

adrianherrera committed:
optimin: fix default timeout

So that it matches the help text

yuawn committed:
remove unused code

vanhauser-thc committed:
update grammar mutator

vanhauser-thc committed:
Merge pull request #1043 from yuawn/fix_MOpt

Remove unused code in MOpt

vanhauser-thc committed:
Merge pull request #1042 from yuawn/fix_afl-cc

Remove redundant if statement in afl-cc

adrianherrera committed:
optimin: support afl-cmin crash mode

vanhauser-thc committed:
fix gramatron

Commit:
Added test for fuzzing a lib on OSX

Commit:
Update readme, minor tweaks

vanhauser-thc committed:
Merge pull request #1044 from WorksButNotTested/osx-lib

Osx lib

jhertz-apple committed:
fix check again clang asan lib to work on mac and linux by removing a character

adrianherrera committed:
optimin: run showmap with an entire input directory

Rather than executing with one seed at a time. This required disabling
the target binary test (for now).

adrianherrera committed:
optimin: removed progress bar

It was hard to get working with showmap's directory mode

Commit:
Changes to add additional FASAN configurations to osx-lib

vanhauser-thc committed:
Merge pull request #1045 from jhertz/fix_asan_check

fix check again clang asan lib to work on mac and linux

vanhauser-thc committed:
Merge pull request #1046 from WorksButNotTested/osx-lib

Changes to add additional FASAN configurations to osx-lib

vanhauser-thc committed:
better macos sysctl settings

adrianherrera committed:
optimin: add target test back in

vanhauser-thc committed:
Merge pull request #1047 from adrianherrera/fixes/optimin

Fixes/optimin

yuawn committed:
fix format string

vanhauser-thc committed:
Merge pull request #1049 from yuawn/dev

fix format string

vanhauser-thc committed:
frida tool fix

vanhauser-thc committed:
frida tool fix

vanhauser-thc committed:
add more string functions for dictionary features

vanhauser-thc committed:
fix testcase

fbeqv committed:
improved error message

vanhauser-thc committed:
makefile fixes

vanhauser-thc committed:
Merge pull request #1050 from fbeqv/better_macos_getshm_errors

improved error message

yuawn committed:
use HASH_CONST

vanhauser-thc committed:
Merge pull request #1052 from yuawn/dev

use HASH_CONST

vanhauser-thc committed:
frida: switch truncate to dd

vanhauser-thc committed:
fix afl_preload issues on macos

DMaroo committed:
Wrote a simple and minimal UI in GTK, with GNUplot graphs embedded in it

TODO: Improve upon the extremely minimal (yet functional) UI

DMaroo committed:
Updated the afl-plot shell script to use GTK UI

TODO: Currently the script is very huge, and needs to be made more
concise and clean. In it's current state, it is hard to maintain.

DMaroo committed:
Added usage info and command line help flag support

TODO: Prevent standalone execution

DMaroo committed:
Added Makefile for building and installing afl-plot-ui

Currently, it offers make and make clean functionality only

DMaroo committed:
Added README to describe the working of afl-plot-ui

DMaroo committed:
Added support for graphical command line flag, improved the general organization of script

Script works as "intended"

DMaroo committed:
Added build and install instructions for afl-plot-ui

I am implicitly assuming that the user already has GK and pkg-config
installed, as requested in the build instructions in the main README

DMaroo committed:
Added afl-plot-ui executable in .gitignore list

DMaroo committed:
Updated README to contain description regarding plot_ui

DMaroo committed:
Added additional dependencies which need to be installed

DMaroo committed:
Updated the CI configuration to install the new dependencies introduced

DMaroo committed:
Changed parameter order, reduced script execution time

DMaroo committed:
Commented out all the builds of afl-plot-ui

Also add the "-" qualifier to denote a possibly unsuccessful build

DMaroo committed:
Fixed a typo

DMaroo committed:
Removed the optional build dependencies for afl-plot-ui

DMaroo committed:
Added separate build and install instructions for afl-plot-ui

vanhauser-thc committed:
Merge pull request #1053 from DMaroo/gtk-dev

Added a minimal working/functional GTK UI for analyzing fuzzing stats

DMaroo committed:
Start plot UI window maximized on launch

DMaroo committed:
Updated README with build and install instructions

vanhauser-thc committed:
Merge pull request #1055 from DMaroo/fullscreen

`afl-plot-ui` window starts maximized now

llzmb committed:
Clean up docs folder (#1059)

Changes:
- Move files to docs/resources.
- Fix references.
- Delete unused files.

llzmb committed:
Change afl to AFL in *.md (#1057)

Changes in *.md files:
- afl++ > AFL++ (acronym)
- afl > AFL (compare https://github.com/google/AFL)

Excluded from changes:
- source code
- function names
- paths (folder and file names)
- URLs

DMaroo committed:
Added resizable panes instead of static separators in afl-plot-ui

The different plots can now be resized using dragging the panes.
The resizing isn't completely smooth from the visual POV (I'm guessing
this is because of embedding GNUplot), but once the position of the
panels are set, the rest of the visual experience is as smooth as it was
before.

DMaroo committed:
Removed HEIGHT and WIDTH since the windows will anyways start maximized

vanhauser-thc committed:
Merge pull request #1060 from DMaroo/gtk-pane

Added resizable panes instead of static separators in afl-plot-ui

llzmb committed:
Clean up references to notes_for_asan.md

ASBOSNIA committed:
Update afl-fuzz-python.c

Fix havoc_mutations not working in python version

vanhauser-thc committed:
Merge pull request #1063 from wxyxsx/stable

Fix "havoc_mutations" not working in Python module

vanhauser-thc committed:
remove outdated references

mqf20 committed:
Warn if "paths_total" property from stats file is inconsistent

vanhauser-thc committed:
Merge pull request #1064 from mqf20/warn_afl_fuzz_stats_paths_total

Warn if `paths_total` property from `fuzzer_stats` file is inconsistent

vanhauser-thc committed:
Merge pull request #1062 from llzmb/docs_cleanup_references_notes_for_asan

Clean up references to notes_for_asan.md

hexcoder- committed:
add AFL++ tutorials from Antonio Morales

vanhauser-thc committed:
add exercise link

vanhauser-thc committed:
AFL_IGNORE_PROBLEMS + library checks and documentation

vanhauser-thc committed:
remove duplicate link

invalid-email-address committed:
Don't print loads of debugging info about cmplog when it isn't enabled

invalid-email-address committed:
Added documetation on how to debug issues

invalid-email-address committed:
Fix to setting of entry_reached

invalid-email-address committed:
Changes to remove hard-coding of MAP_SIZE_POW2 when inline instrumentation is disabled

vanhauser-thc committed:
Merge pull request #1065 from WorksButNotTested/fixes

Fixes

b1gr3db committed:
Fixed spelling of quarantine

vanhauser-thc committed:
Merge pull request #1066 from b1gr3db/dev

Fixed spelling of quarantine

vanhauser-thc committed:
uninstall makefile target

vanhauser-thc committed:
fix typo

invalid-email-address committed:
Add support for generating coverage information

invalid-email-address committed:
Changes to build many-linux build from working copy instead of a fixed branch

invalid-email-address committed:
Fix coverage data

vanhauser-thc committed:
switch back to -j4 for building llvm

vanhauser-thc committed:
Merge pull request #1071 from WorksButNotTested/many-local

Many local

invalid-email-address committed:
Force exit of coverage process

vanhauser-thc committed:
Merge pull request #1073 from WorksButNotTested/coverage2

Force exit of coverage process

vanhauser-thc committed:
more partial linking

vanhauser-thc committed:
fix shared linking on macos

domenukk committed:
update unicornafl

hexcoder- commented on frida_mode/DEBUGGING.md:
```diff
@@ -0,0 +1,160 @@
+generated by FRIDA is likely to be very difficult to follow. For this reason,
+the following debugging strategies are outlined below.
+
+Byte convention below all files should be provided with their path (they are
```
"By convention" instead of "Byte convention"?

vanhauser-thc replied:
@WorksButNotTested

WorksButNotTested replied:
Yep. Typo sorry.

hexcoder- replied:
fixed, Thanks.

hexcoder- committed:
typos

hexcoder- committed:
wording

hexcoder- committed:
typos / wording

hexcoder- commented on instrumentation/README.persistent_mode.md:
```diff
@@ -2,13 +2,16 @@
 
-The most effective way is to fuzz in persistent mode, as the speed can easily
+In persistent mode, AFL++ fuzzes a target multiple times
+in a single process, instead of forking a new process for each fuzz execution.
```
This is technically unprecise, I think, as the forkserver already avoided fork()ing and used clone() and COW. 
The main points here (for me) is: 
 - the life time of the target is extended over multiple fuzzing runs.
 - The target implements a loop and runs like a nonterminating server waiting for inputs.
 - AFL supplies these inputs.

vanhauser-thc replied:
I tried to clarify it a little bit better. I am not sure that your main points make it easier to understand either, could work for some and not for others. I think this only works with seeing example code, whatever the writing.

hexcoder- replied:
Yes, I agree; lets delegate this to an example.

hexcoder- commented on instrumentation/SanitizerCoverageLTO.so.cc:
```diff
@@ -626,12 +626,41 @@ bool ModuleSanitizerCoverage::instrumentModule(
+                              !FuncName.compare("xmlStrncasecmp") ||
+                              !FuncName.compare("g_ascii_strncasecmp") ||
+                              !FuncName.compare("Curl_strncasecompare") ||
+                              !FuncName.compare("g_strncasecmp"));
```
If these lists keep on growing, we could soon have namespace issues (clashes with user functions named the same).

vanhauser-thc replied:
that is not an issue IMHO. it is not only matched on function name but also parameter types and return type. and even if these all match and still it does something completely different the impact is minimal

hexcoder- replied:
ok with me. Lets see, if someone experiences this conflict at all.

hexcoder- committed:
fix typo fix

hexcoder- committed:
add newline at end

hexcoder- committed:
wording

hexcoder- reviewed (approved):
LGTM, thanks.

vanhauser-thc committed:
revise paragraph

hexcoder- committed:
typo

invalid-email-address committed:
Fixed coverage on OSX (dependency on pipe2)
Removed use of 'realpath' in makefiles to fix OSX incompatibility
Fixed handling of when prefetching should be enabled
Snap the main binary during initialization to avoid stability issues with lazy loading
Add support for configurable inline cache entries for FRIDA on x86/x64
Support for prefetching FRIDA backpatches on x86/x64
Improved stats support on x86/x64/aarch64

vanhauser-thc committed:
Merge pull request #1075 from WorksButNotTested/test

Various New Features & Fixes
